builtin/aws/lambda/platform.go

```diff
@@ -269,39 +277,65 @@ func (p *Platform) Deploy(
 			reset = true
 		}
 
+		if *curFunc.Configuration.EphemeralStorage.Size != storage {
+			update.EphemeralStorage = &lambda.EphemeralStorage{
+				Size: aws.Int64(storage),
+			}
+			reset = true
+		}
+
 		if reset {
+			step.Update("Detected a configuration change. Updating Lambda function...")
+
 			update.FunctionName = curFunc.Configuration.FunctionArn
 
 			_, err = lamSvc.UpdateFunctionConfiguration(&update)
 			if err != nil {
 				return nil, errors.Wrapf(err, "unable to update function configuration")
 			}
+
+			step.Update("Updated Lambda configuration.")
 		}
 
-		funcCfg, err := lamSvc.UpdateFunctionCode(&lambda.UpdateFunctionCodeInput{
-			FunctionName:  aws.String(src.App),
-			ImageUri:      aws.String(img.Name()),
-			Architectures: aws.StringSlice([]string{architecture}),
-		})
+		step.Done()
 
-		if err != nil {
-			if aerr, ok := err.(awserr.Error); ok {
-				switch aerr.Code() {
-				case "ValidationException":
-					// likely here if Architectures was invalid
-					if architecture != lambda.ArchitectureX8664 && architecture != lambda.ArchitectureArm64 {
-						return nil, fmt.Errorf("architecture must be either \"x86_64\" or \"arm64\"")
+		// the following update has retry behavior in case the
+		// previous `UpdateFunctionConfiguration` call occurs
+		step = sg.Add("Updating function code...")
+		for i := 0; i < 30; i++ {
+			funcCfg, err := lamSvc.UpdateFunctionCode(&lambda.UpdateFunctionCodeInput{
+				FunctionName:  aws.String(src.App),
+				ImageUri:      aws.String(img.Name()),
+				Architectures: aws.StringSlice([]string{architecture}),
+			})
+
+			if err == nil {
+				funcarn = *funcCfg.FunctionArn
+				step.Update("Updated function code!")
+				break
+			}
+
+			if err != nil {
+
+				if aerr, ok := err.(awserr.Error); ok {
+					switch aerr.Code() {
+					// unrecoverable error
+					case "ValidationException":
+						// likely here if Architectures was invalid
+						if architecture != lambda.ArchitectureX8664 && architecture != lambda.ArchitectureArm64 {
+							return nil, fmt.Errorf("architecture must be either \"x86_64\" or \"arm64\"")
+						}
+						// other validation error
+						return nil, err
+					// Function update in progress error
+					case "ResourceConflictException":
+						log.Warn(fmt.Sprintf("Function update in progress. Waiting 2 seconds before retrying... (%d/30)", i+1))
+						time.Sleep(2 * time.Second)
```

Comment:
This timeout makes sense, but let's leverage `context` instead of a sleep and for iterator. You can see an example of how this might work in the nomad jobspec releaser:

https://github.com/hashicorp/waypoint/blob/8d9e228fb8506ab7bfb169ca778b9b6dc206ec36/builtin/nomad/jobspec/releaser.go#L176-L178

https://github.com/hashicorp/waypoint/blob/8d9e228fb8506ab7bfb169ca778b9b6dc206ec36/builtin/nomad/jobspec/releaser.go#L184-L188

Let me know if you have any questions or need help figuring this out!

Reply:
Neat and TIL! Addressed - https://github.com/hashicorp/waypoint/pull/3213/commits/d046cbccab1ef7217e12fcdabc671ec4400f9fde